src/V3Ast.h

```diff
@@ -1247,6 +1247,45 @@ class VBasicTypeKey final {
     ~VBasicTypeKey() = default;
 };
 
+// ######################################################################
+//  VSelfPointerText - Represents text to be emitted before a given var reference, call, etc. to
+//  serve as a pointer to a 'self' object. For example, it could be empty (no self pointer), or the
+//  string 'this', or 'vlSymsp->...'
+
+class VSelfPointerText final {
+private:
+    // STATIC MEMBERS
+    // Keep these in shared pointers to avoid branching for special cases
+    static const std::shared_ptr<const string> s_emptyp;
+    static const std::shared_ptr<const string> s_thisp;
+
+    // MEMBERS
+    std::shared_ptr<const string> m_strp;
+
+    // CONSTRUCTORS
+    VSelfPointerText(std::shared_ptr<const string> strp)
+        : m_strp{strp} {}
+    VSelfPointerText(string&& str)
+        : m_strp{std::make_shared<const string>(std::move(str))} {}
+
+public:
+    // METHODS
+    bool isEmpty() const { return m_strp == s_emptyp; }
+    bool isVlSym() const { return m_strp->find("vlSymsp") != string::npos; }
+    bool hasThis() const { return m_strp == s_thisp || VString::startsWith(*m_strp, "this"); }
+    string protect(bool useSelfForThis, bool protect) const;
+    const std::string& asString() const { return *m_strp; }
+    bool operator==(const VSelfPointerText& other) const { return *m_strp == *other.m_strp; }
+
+    // STATIC METHODS
+    static VSelfPointerText makeEmpty() { return {s_emptyp}; }
+    static VSelfPointerText makeThis() { return {s_thisp}; }
+    static VSelfPointerText makeThis(const string& field) { return {"this->" + field}; }
+    static VSelfPointerText makeVlSyms(const string& field) {
+        return {"(&vlSymsp->" + field + ')'};
+    }
```

Comment:
I guess this is OK, but these really seem like constructors, and in most other places we use a class to select a specialized constructor. E.g. see AstConst. Then would call like this

     VlSelfPointerText{VlSelfPointerText::Empty{}}
     VlSelfPointerText{VlSelfPointerText::This{}}
     VlSelfPointerText{VlSelfPointerText::This{}, field}
     VlSelfPointerText{VlSelfPointerText::VlSyms{}, field}

(see comment below)

But I don't feel strongly about this.

Reply:
I think it makes sense, switched to type tags.